R/data.R

```diff
@@ -123,69 +123,96 @@ any_na_elements <- function(data) {
 }
 
 
+draws_to_csv <- function(draws, sampler_diagnostics = NULL) {
+  sampler_diagnostic_names <- c("accept_stat__", "stepsize__", "treedepth__", "n_leapfrog__", "divergent__", "energy__")
+
+  n <- posterior::niterations(draws)
+  n_chains <- posterior::nchains(draws)
+  if (is.null(sampler_diagnostics)) {
+    # create dummy sampler diagnostics due to CmdStan requirement for all columns in GQ
+    sampler_diagnostics <- rep(0, n * length(sampler_diagnostic_names) * n_chains)
+    dim(sampler_diagnostics) <- c(n, n_chains, length(sampler_diagnostic_names))
+    sampler_diagnostics <- posterior::as_draws_array(sampler_diagnostics)
+    posterior::variables(sampler_diagnostics) <- sampler_diagnostic_names
+  }
+
+  # the columns must be in order "lp__, sampler_diagnostics, parameters"
+  variables <- posterior::variables(draws)
+  # create a dummy lp__ column if it does not exist
+  if ("lp__" %in% variables) {
+    lp__ <- NULL
+  } else {
+    lp__ <- rep(0, n * n_chains)
+    dim(lp__) <- c(n, n_chains, 1)
+    lp__ <- posterior::as_draws_array(lp__)
+    posterior::variables(lp__) <- "lp__"
+  }
```

Comment:
This looks good, but I think there's a slightly nicer way code it using the `posterior::draws_array()` constructor. I can do that in a few minutes and push it.

Reply:
That would be fantastic! Thanks!

Reply:
Just pushed it. Let me know if you like it. If not we can revert because what you had definitely also worked.

Reply:
Yes, I like it.